provider/datagen/src/driver.rs

```diff
@@ -954,84 +1012,118 @@ fn select_locales_for_key(
+        let selected_locales = supported_map.into_values().flatten().collect();
+        return Ok(selected_locales);
     }
 
+    // Need the fallbacker now.
     let fallbacker = fallbacker.as_ref().map_err(|e| *e)?;
     let fallbacker_with_config = fallbacker.for_config(key.fallback_config());
 
-    // TODO(#4629): Allow the exclusion of und
-    let include_und = true;
-
-    let result = supported_and_explicit
-        .filter(|locale_orig| {
-            let mut locale = locale_orig.clone();
-            locale.remove_aux();
-            if implicit.contains(&locale) {
-                return true;
-            }
-            if explicit.contains(&locale) {
-                return true;
+    // Compute a map from LanguageIdentifiers to DataLocales, including inherited auxiliary keys
+    // and extensions. Also resolve the ancestors and descendants while building this map.
+    let mut selected_langids = requested_families.keys().cloned().collect::<HashSet<_>>();
+    let expansion_map: HashMap<&LanguageIdentifier, HashSet<DataLocale>> = supported_map
+        .keys()
+        .chain(requested_families.keys())
+        .map(|current_langid| {
+            let mut expansion = supported_map
+                .get(current_langid)
+                .cloned()
+                .unwrap_or_default();
+            if include_full && !selected_langids.contains(current_langid) {
+                log::trace!("Including {current_langid}: full locale family: {key}");
+                selected_langids.insert(current_langid.clone());
             }
-            if locale.is_langid_und() && include_und {
-                return true;
-            }
-            if locale.language().is_empty()
-                && matches!(
-                    key.fallback_config().priority,
-                    icu_provider::FallbackPriority::Region
-                )
-            {
-                return true;
+            if current_langid.language.is_empty() && current_langid != &LanguageIdentifier::UND {
+                log::trace!("Including {current_langid}: und variant: {key}");
+                selected_langids.insert(current_langid.clone());
             }
-            // Special case: skeletons *require* the -u-ca keyword, so don't export locales that don't have it
-            // This would get caught later on, but it makes datagen faster and quieter to catch it here
-            if key.path().get() == "datetime/skeletons@1" && !locale.has_unicode_ext() {
-                return false;
-            }
-            let mut iter = fallbacker_with_config.fallback_for(locale);
-            while !iter.get().is_und() {
-                if explicit.contains(iter.get()) {
-                    return true;
+            let include_ancestors: bool = requested_families
+                .get(current_langid)
+                .map(|family| family.include_ancestors)
+                .unwrap_or(false);
```

Comment:
please document when this `false` will be used. iiuc it's when a not-requested but supported locale is generated?

Reply:
Added comments.